integration/src/main/java/scripts/testscripts/WorkspacePermissionOperations.java

```diff
@@ -0,0 +1,173 @@
+
+  private TestUserSpecification adminUser;
+  private TestUserSpecification regularUser;
+
+  private RawlsClient adminRawlsClient;
+  private RawlsClient userRawlsClient;
+  private WorkspaceDetails adminTestWorkspace;
+  private WorkspaceDetails userTestWorkspace;
+
+  // Catalog APis
+  private final List<UUID> datasetIds = new ArrayList<>();
+  private DatasetsApi adminDatasetsApi;
+  private DatasetsApi userDatasetsApi;
+  private UUID adminTestDatasetId;
+
+  @Override
+  public void setup(List<TestUserSpecification> testUsers) throws Exception {
+    for (TestUserSpecification testUser : testUsers) {
+      if (testUser.userEmail.equals(ADMIN_EMAIL)) {
+        adminUser = testUser;
+      }
+      if (testUser.userEmail.equals(USER_EMAIL)) {
+        regularUser = testUser;
+      }
+    }
+    assertNotNull(adminUser);
+    assertNotNull(regularUser);
+
+    adminRawlsClient = new RawlsClient(server, adminUser);
+    userRawlsClient = new RawlsClient(server, regularUser);
+    adminDatasetsApi = new DatasetsApi(new CatalogClient(server, adminUser));
+    userDatasetsApi = new DatasetsApi(new CatalogClient(server, regularUser));
+    adminTestWorkspace = adminRawlsClient.createTestWorkspace();
+    userTestWorkspace = userRawlsClient.createTestWorkspace();
+    adminTestDatasetId =
+        adminCreateDataset(datasetRequestForWorkspace(adminTestWorkspace.getWorkspaceId()));
+  }
+
+  @Override
+  public void userJourney(TestUserSpecification testUser) throws Exception {
+    testReaderPermissions();
+    testNoPermissions();
+    testAdminPermissionsOnUserSnapshot();
+  }
+
+  private void testReaderPermissions() throws Exception {
+    setTestWorkspacePermissionForRegularUser(WorkspaceAccessLevel.READER.getValue());
+    // User cannot create a catalog entry on snapshot when user only has reader access on snapshot
+    CreateDatasetRequest request = datasetRequestForWorkspace(adminTestWorkspace.getWorkspaceId());
+    // note if this assertion fails we'll leave behind a stray dataset in the db
+    assertThrows(ApiException.class, () -> userDatasetsApi.createDataset(request));
+    assertThat(
+        userDatasetsApi.getApiClient().getStatusCode(),
+        is(HttpStatusCodes.STATUS_CODE_UNAUTHORIZED));
+
+    // verify the user can still access data though
+    userDatasetsApi.getDataset(adminTestDatasetId);
+    assertThat(userDatasetsApi.getApiClient().getStatusCode(), is(HttpStatusCodes.STATUS_CODE_OK));
+  }
+
+  private void testNoPermissions() throws Exception {
+    clearTestWorkspacePermissions();
+    // verify the user cannot access dataset without any permissions
+    assertThrows(ApiException.class, () -> userDatasetsApi.getDataset(adminTestDatasetId));
+    assertThat(
+        userDatasetsApi.getApiClient().getStatusCode(), is(HttpStatusCodes.STATUS_CODE_NOT_FOUND));
+  }
+  // TODO (DC-446): Fix with https://broadworkbench.atlassian.net/browse/DC-446
```

Comment:
Since there's a JIRA for this, can this test code go into the JIRA as well? I think that would be better than leaving commented out code in the source.

Reply:
That sounds great! I dropped the code into a comment on the ticket.